miking-lang/miking
# Parse options for subcommands

* Handle parsing of options for subcommands. For example, running `mi --help` shows a main menu like before and `mi tune --help` shows additional options valid for `tune`:
```
Usage: mi tune [<options>] file [<options>]

Options:
  --debug-parse            Print the AST after parsing
  --debug-generate         Print the AST after code generation
  --debug-type-annot       Print the AST after adding type annotations
  --debug-profile          Instrument profiling expressions to AST
  --exit-before            Exit before evaluation or compilation
  --disable-prune-utests   Disable pruning of utests with missing external 
                           dependencies
  --disable-prune-warning  Disable warning when pruning utests with missing 
                           external dependencies
  --test                   Generate utest code
  --disable-optimizations  Disables optimizations to decrease compilation time
  --tuned                  Use tuned values when compiling, or as defaults when 
                           tuning
  --compile                Compile directly after tuning
  --cpu-only               Translate accelerated code to multicore CPU code
  --typecheck              Type check the program before evaluation or 
                           compilation
  --output <file>          Write output to <file> when compiling
  --help                   Display this list of options
  --verbose                Print the search state during tuning
  --iters <n>              Maximum number of iterations to perform before 
                           exiting the search
  --timeout-ms <ms>        Maximum time in ms to conduct the tuning
  --warmups <n>            Number of warmup iterations before starting tuning
  --method <method>        Search method, one of {simulated-annealing, 
                           tabu-search, random-walk, exhaustive, 
                           semi-exhaustive, binary-search}
  --input <s>              Input data to the tuned program. Note that several 
                           inputs can be given by providing several --input
  --sa-init-temp <t>       If --method simulated-annealing is used, this gives 
                           the initial temperature (default 100.)
  --sa-decay-factor <d>    If --method simulated-annealing is used, this gives 
                           the decay factor (default 0.95)
  --tabu-size <n>          If --method tabu-search is used, this gives the 
                           number of configurations to keep in the tabu list 
                           (default 10)
  --step-size <n>          Step size for integer ranges (default 1)
  --ignore-errors          Ignore errors during tuning
  --disable-exit-early     Let the process run to completion during tuning
  --seed <n>               Set the seed for random search
  --enable-cleanup         Clean up tune file after tuning
  -- <args>                If the run or eval commands are used, then the texts
                           following -- are arguments to the executed program
```
I haven’t split up the options for the other subcommands yet but it should be easy now.
* Clean up `mexpr.prof` file and `.tune` files when running the tests.
* Remove a dependency on the `timeout` command when running `make test-tune`
* Add an `argToFloatMin` function to argument parser
* Remove `-j` flag from GitHub workflow file to remove annoying error messages
